### PR TITLE
Refactor dip25 error/deprecation handling code

### PR DESCRIPTION
**Pull request by atilaneves**
DIP25 has been technically enabled by default by causing the compiler to emit deprecation warnings if the code wouldn't compile with `-preview=dip25` but in an ad-hoc fashion that would make it hard to repeat the process for other preview switches. With an eye towards enabling dip1000, this PR refactors the existing code.

### CODE REVIEW

File: src/dmd/errors.d [hunk 1]
```diff
@@ -35,6 +35,37 @@ enum Classification : Color
...
 
+
+enum AlwaysError
+{
+    no,
+    yes,
+}
+
+
+pragma(printf) private extern (C++) void noop(const ref Loc loc, const(char)* format, ...) {}
```
Comment (thewilsonator) [inline src/dmd/errors.d:46]: note the `static if (__VERSION__ < 2092)` version guards for `pragma(printf)` below

File: .gitignore [hunk 2]
```diff
@@ -12,6 +12,7 @@ generated/
 -trace.log
 *.lst
 .dub
+dub_describe.json
```
Comment (Geod24) [inline .gitignore:15]: What's that ? It's not output by `dub` and I couldn't find a reference in this repository.

File: test/compilable/bug21196.d [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
-REQUIRED_ARGS: -de
+REQUIRED_ARGS: -de -revert=dip25
```
Comment (thewilsonator) [inline test/compilable/bug21196.d:2]: this seems unrelated

File: .gitignore (hunk 2 above)
Comment (thewilsonator) [inline .gitignore:15]: as does this

**Review by Geod24 — DISMISSED** [review]
Smart. Just a comment about the first commit. (EDIT: And the `-revert`)

File: test/compilable/bug21196.d (hunk 3 above)
Comment (atilaneves) [reply test/compilable/bug21196.d:2]: Nope. Otherwise this fails.

File: .gitignore (hunk 2 above)
Comment (atilaneves) [reply .gitignore:15]: I don't know. I'll remove it from the PR though since it's unrelated. I do know that it keeps popping up and has started doing so lately. For me anyway.

File: test/compilable/bug21196.d (hunk 3 above)
Comment (thewilsonator) [reply test/compilable/bug21196.d:2]: Then this isn't a refactoring.

File: src/dmd/errors.d (hunk 1 above)
Comment (atilaneves) [reply src/dmd/errors.d:46]: Ah. I reckoned they weren't needed for newer versions of the compiler but I forgot about bootstrapping.

Comment (thewilsonator) [reply src/dmd/errors.d:46]: Correct

File: test/compilable/bug21196.d (hunk 3 above)
Comment (atilaneves) [reply test/compilable/bug21196.d:2]: I should explain better. The code before (which made this test pass) was:

```d
             if (emitError)
                    error(e.loc, msg, e.toChars(), kind, v.toChars());
                else if (!sc.isDeprecated())
                    deprecation(e.loc, msg, e.toChars(), kind, v.toChars());
```

Notice that the deprecation was only printed in the case of `!sc.isDeprecated`. That was an ad-hoc solution to https://issues.dlang.org/show_bug.cgi?id=21196 whereas the real way to get rid of deprecation warnings like the ones here is to pass `-revert=previewXXX`. That's why `noop` exists now.

File: .gitignore (hunk 2 above)
Comment (atilaneves) [reply .gitignore]: Found out where - it's from my flycheck-dmd-dub Emacs package which got a PR merged recently.

File: test/compilable/bug21196.d (hunk 3 above)
Comment (Geod24) [reply test/compilable/bug21196.d:2]: Hold on, does it mean that with this change, deprecations will show up in deprecated scopes ?

File: .gitignore (hunk 2 above)
Comment (Geod24) [reply .gitignore]: You might want https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/

**Comment by RazvanN7** [comment]
The C++ headers need to be updated. Signatures should be added in `errors.h` and `frontend.h`.

**Comment by atilaneves** [comment]
> The C++ headers need to be updated. Signatures should be added in `errors.h` and `frontend.h` for the newly added functions in `errors.d`.

Does `errors.h` really need to be updated? Who would be calling these new functions from C++?

File: src/dmd/errors.d [hunk 4]
```diff
@@ -35,6 +35,41 @@ enum Classification : Color
     tip = Color.brightGreen,          /// for tip messages
 }
 
+
+enum AlwaysError
+{
+    no,
+    yes,
+}
```
Comment (thewilsonator) [inline src/dmd/errors.d:43]: this appears to be unused

Comment (atilaneves) [reply src/dmd/errors.d:43]: Argh, leftover from previous experiments. Good catch.

File: src/dmd/frontend.h [hunk 5]
```diff
@@ -7985,4 +7980,3 @@ extern "C" Object* _d_newclass(const TypeInfo_Class* const ci);
 extern "C" void* _d_newitemT(TypeInfo* ti);
 
 extern "C" void* _d_newitemiT(TypeInfo* ti);
-
```
Comment (thewilsonator) [inline src/dmd/frontend.h:7982]: you removed this too

Comment (atilaneves) [reply src/dmd/frontend.h:7982]: That's... an empty line. But ok.

**Comment by atilaneves** [comment]
> The C++ headers need to be updated. Signatures should be added in `errors.h` and `frontend.h` for the newly added functions in `errors.d`.

I changed their visibility to `package`.

**Comment by atilaneves** [comment]
@Geod24 Should be good to go.

**Comment by Geod24** [comment]
What about:
https://github.com/dlang/dmd/pull/12426#discussion_r611497938

File: test/compilable/bug21196.d (hunk 3 above)
Comment (atilaneves) [reply test/compilable/bug21196.d:2]: > Hold on, does it mean that with this change, deprecations will show up in deprecated scopes ?

Example?

If you mean that, in the original issue you filed, that they'll show up, then yes, *unless* `-revert=preview25` is used.

Comment (Geod24) [reply test/compilable/bug21196.d:2]: Then it's introducing a regression. Having deprecations, *even language deprecations*, be silenced when in a `deprecated` scope is an important feature, as it allows libraries to transition away from one API to the other.
The use case that prompted the issue was a library that would have an overload which internally kept pointers to stack allocated data in `void[]` slices. Transitioning to the other, safer overload was trivial, so in the library we deprecated the function that relied on the non-dip25 behavior. The other alternative would be either have everyone see an unrelated deprecation showing up on each build, or force us to make a breaking change outside of our release cycle, none of which is acceptable.

Comment (Geod24) [reply test/compilable/bug21196.d:2]: > Example?

This very test case. Both functions are deprecated and aren't called from a deprecated scope, hence no deprecation should be issued.

Comment (atilaneves) [reply test/compilable/bug21196.d:2]: I clearly didn't understand that issue or the test for it. Fixing it now.

**Comment by atilaneves** [comment]
@Geod24 I think I addressed the issues now.

**Comment by thewilsonator** [comment]
```
../dmd/src/dmd/doc.d(364,34): Error: file `"default_ddoc_theme.ddoc"` cannot be found or not in a path specified with `-J`
```

rebase?

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Hmm, still there. IDK why that fails.

**Comment by MoonlightSentinel** [comment]
#12368 didn't add the import path to `dub.sdl` => #12436

**Comment by atilaneves** [comment]
Ugh, now a seemingly random ldc failure?

**Comment by atilaneves** [comment]
I ran the semaphore script on my machine with `DMD=ldc` and... it didn't fail.

**Comment by atilaneves** [comment]
semaphore is my new least favourite thing, especially when, right now, I can't get the script to pass on my machine on *master*, let alone reproduce this failure.

**Comment by MoonlightSentinel** [comment]
FYI: We've disabled the join_detach test (that caused the semaphore failures) a few hours ago, so that should pass in a rerun.

**Comment by atilaneves** [comment]
Green on semaphore!

Do I need a new approval or can I slap an automerge tag on it?

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
For the sake of bureaucracy I did it for you.
